File: topics/BackTrack/N_queen.java
package BackTrack;

import java.util.Scanner;

public class N_queen {
	
	static int[][] board;
	static int n;
	static boolean finished;
	public static void main(String args[])
	{
		Scanner in = new Scanner(System.in);
		n = in.nextInt();
		in.close();
		board = new int[n][n];
		finished = false;
		backTrack(0);
		
		StringBuilder result = new StringBuilder();
		for(int i = 0; i < n; i++)
		{
			for(int j = 0; j < n; j++)
			{
				if(board[i][j] == 0)
				{
					result.append(".");
				}
				else
				{
					result.append("x");
				}
			}
			result.append("\n");
		}
		
		System.out.println(result);
	}
	
	private static void backTrack(int col)
	{
		if(col >= n)
		{
			return;
		}
		else
		{
			for(int i = 0; i < n; i++)
			{
				if(!dangerous(i,col))
				{
					board[i][col] = 1;
					if(col == n - 1)
					{
						finished = true;
						return;
					}
					backTrack(col + 1);
				}
				
				if(finished)
				{
					break;
				}
				else
				{
					board[i][col] = 0;
				}
			}
		}
	}
	
	private static boolean dangerous(int row, int col)
	{
		for(int i = 0; i < col; i++)
		{
			if(board[row][i] == 1)
			{
				return true;
			}
		}
		
		int x = row - 1;
		int y = col - 1;
		while(x >= 0 && y >= 0)
		{
			if(board[x--][y--] == 1)
			{
				return true;
			}
		}
		
		int i = row + 1;
		int j = col - 1;
		while(i < n && j >= 0)
		{
			if(board[i++][j--] == 1)
			{
				return true;
			}
		}
		
		return false;
		
	}

}

File: topics/BackTrack/N_queen.py
#https://pcs.cs.cloud.vt.edu/contests/36/problems/A
class point:
    def __init__(self, row, col):
        self.row = row
        self.col = col
def in_range(n, p):
    return p.row >= 0 and p.row < n and p.col >= 0 and p.col < n
def valid(board, n, row, col):
    for i in range(n):
        if i != col:
            if board[row][i] == 1:
                return False
    for i in range(n):
        if i != row:
            if board[i][col] == 1:
                return False
    x1 = point(row-1,col-1)
    while in_range(n,x1):
        if board[x1.row][x1.col] == 1:
            return False
        x1.row -= 1
        x1.col -= 1
    x2 = point(row-1,col+1)
    while in_range(n,x2):
        if board[x2.row][x2.col] == 1:
            return False
        x2.row -= 1
        x2.col += 1
    x3 = point(row+1,col+1)
    while in_range(n,x3):
        if board[x3.row][x3.col] == 1:
            return False
        x3.row += 1
        x3.col += 1
    x4 = point(row+1,col-1)
    while in_range(n,x4):
        if board[x4.row][x4.col] == 1:
            return False
        x4.row += 1
        x4.col -= 1
    return True

n = int(input())
board = [[0 for i in range(n)] for j in range(n)]
finished = False

def backtrack(board, col):
    global finished
    if finished:
        return
    for row in range(n):
        if valid(board,n,row,col):
            board[row][col] = 1
            if col == n-1:
                finished = True
            backtrack(board, col+1)
            if finished:
                return
            else:
                board[row][col] = 0
                
backtrack(board, 0)
for i in range(n):
    temp = ""
    for j in range(n):
        if board[i][j] == 0:
            temp += "."
        else:
            temp += "x"
    print(temp)